Source/WebCore/Modules/compression/DecompressionStreamDecoder.h

```diff
@@ -60,18 +72,39 @@ class DecompressionStreamDecoder : public RefCounted<DecompressionStreamDecoder>
     const size_t startingAllocationSize = 16384; // 16KB
     const size_t maxAllocationSize = 1073741824; // 1GB
     
-    bool initailized { false };
-    bool finish { false };
-    z_stream zstream;
+    bool m_initailized { false };
+    bool m_finish { false };
+    z_stream m_zstream;
+
+    bool m_usingAppleCompressionFramework { false };
+
+#if PLATFORM(COCOA)
+    compression_stream m_stream;
+    ExceptionOr<RefPtr<JSC::ArrayBuffer>> decompressAppleCompressionFramework(const uint8_t* input, const size_t inputLength);
+    ExceptionOr<bool> initializeAppleCompressionFramework();
+#endif
+
+    ExceptionOr<RefPtr<JSC::ArrayBuffer>> (DecompressionStreamDecoder::*decompress)(const uint8_t*, const size_t);
 
     Formats::CompressionFormat m_format;
 
-    ExceptionOr<Vector<uint8_t>> decompress(const uint8_t* input, const size_t inputLength);
+    ExceptionOr<RefPtr<JSC::ArrayBuffer>> decompressZlib(const uint8_t* input, const size_t inputLength);
     ExceptionOr<bool> initialize();
 
     explicit DecompressionStreamDecoder(unsigned char format) 
         : m_format(static_cast<Formats::CompressionFormat>(format))
     {
+        decompress = &DecompressionStreamDecoder::decompressZlib;
```

Comment:
I am not sure it is worth having a function pointer here.
This is totally unneeded for non COCOA ports.
And for Cocoa ports, we could simply introduce a decompress method that would check m_formatto either call decompressAppleCompressionFramework or decompressZlib.
Code would be easier to read and I do not think this will change much in efficiency.

Reply:
I was debating whether to use function pointers or not.

The code in `decode()` and `flush()` got a lot uglier to read without it.

You would ultimately need to do something like below...

```CPP
RefPtr<JSC::ArrayBuffer> compressedDataCheck;

#if PLATFORM(COCOA)
if(m_usingAppleCompressionFramework)
   compressedDataCheck = decompressAppleCompressionFramework(data, input.length());
else
   compressedDataCheck = decompressZlib(data, input.length());
#else
   compressedDataCheck = decompressZlib(data, input.length());
#endif
```

I also did try a few other methods of abstraction, but function pointers came out as the least ugly to read at the end of the day.

Function pointers allow us to hide all the complex logic in the initializer, so `decode()` and `flush()` read cleanly.

I agree with the point the function pointers serve little purpose on non apple platforms.

Reply:
Got rid of the function ptrs in a clean way.